### DISCUSSION TITLE
Guidance on Handling Highly Imbalanced Exoplanet Dataset

### DISCUSSION THREAD

**Discussion by prabej7** [nasa/spaceapps#1573] [Challenges] [answered]
Hello everyone,

I’m working on a machine learning model for exoplanet classification using NASA’s open datasets. My dataset is highly imbalanced: about 80% are CONFIRMED planets, while the remaining 20% are FALSE POSITIVE and CANDIDATE planets.

The positive side is that my model generalizes well on the overall dataset, but I’m struggling with predicting the rare classes accurately.

I’m looking for guidance on techniques or approaches to improve performance on the underrepresented classes without sacrificing overall model generalization. Specifically, I’m interested in:

- Resampling strategies (oversampling/undersampling)
- Any other best practices for highly imbalanced multi-class datasets
- Any advice, references, or examples would be highly appreciated!

Thanks in advance!

**Comment by ivenzor** [comment, marked as answer]
Hi, if you are using boosting or similar methods you can give more weight to the minority class with the parameter `scale_pos_weight`
Check: https://xgboost.readthedocs.io/en/stable/parameter.html
You can also tune min_child_weight, gamma, and max_delta_step along with scale_pos_weight to stabilize training on imbalanced data.
You can use stratified K-fold to ensure all classes are represented in each split. Otherwise, the rare classes may vanish from a validation fold entirely.

If using neural networks or scikit-learn models, specify class weights or compute weights inversely proportional to class frequency.

For resampling there is also SMOTE (Synthetic Minority Over-sampling Technique) and its variants (Borderline-SMOTE, SMOTE-Tomek) can generate synthetic minority samples.

Also when using evaluation metrics avoid accuracy and use more robust metrics instead such as Macro-averaged F1 or Matthews correlation coefficient (MCC)

For exoplanets, false positives often come from eclipsing binaries or systematics. You can try to add astrophysical features (transit depth ratios, odd/even depth comparison, centroid motion) to help the classifier discriminate such cases.

**Comment by srinivas-sme** [comment]
Resampling techniques are mainly to balance the class distribution with respect to the dataset.
Begin with simpler techniques or random oversampling or under sampling to see the immediate impact.

**Comment by arpanghoshal** [comment]
**Short answer:** Treat this as a *cost-sensitive, leakage-aware, multi-class* problem. Prefer class/sample weights + proper metrics/thresholds over heavy resampling. Use astrophysics-aware features and guard against leakage across the same star IDs.

**What’s working & what to fix**

* Accuracy looks good because the majority class dominates. Optimize **macro-F1 / macro-recall / MCC** and **per-class PR-AUC**, not accuracy.
* In multi-class XGBoost, `scale_pos_weight` only applies to binary; instead use **`sample_weight`** (or one-vs-rest).

---

### 1) Make the learner cost-sensitive (multi-class safe)

**Option A: sample weights (recommended):**

```python
import numpy as np
from sklearn.utils.class_weight import compute_class_weight
from xgboost import XGBClassifier

classes = np.unique(y_train)  # e.g., ["CONFIRMED","CANDIDATE","FALSE_POSITIVE"]
cw = compute_class_weight(class_weight="balanced", classes=classes, y=y_train)
class_to_w = {c:w for c,w in zip(classes, cw)}
w_train = np.array([class_to_w[c] for c in y_train])

clf = XGBClassifier(
    objective="multi:softprob",
    eval_metric="mlogloss",
    n_estimators=600,
    max_depth=6,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.8,
    reg_lambda=2.0
)
clf.fit(X_train, y_train, sample_weight=w_train)
```

**Option B: one-vs-rest (OVR):** Train three binary classifiers with class-specific `scale_pos_weight`, then combine calibrated probabilities.

**Neural nets:** use **weighted cross-entropy** or **focal loss** with class weights.

---

### 2) Resampling (use lightly, and only on tabular features)

* Try **random undersampling** of the majority + **light oversampling** of minority.
* **SMOTE/Borderline-SMOTE/SMOTE-Tomek** can help on *derived/tabular* features. Avoid SMOTE directly on raw **time-series light curves** (it distorts astrophysical shape). If you must augment sequences, use domain-safe jittering (small time/flux noise), cutout, or mixup on *feature embeddings*, not raw flux.

---

### 3) Split strategy: kill leakage early

* Use **StratifiedGroupKFold** (group by **star/target ID**; e.g., TIC/Kepler ID) so the same object never appears in both train and val.
* Keep class proportions across folds (**stratified**). Rare classes must appear in every fold.

---

### 4) Metrics & thresholds you actually care about

* Report **macro-F1, macro-recall, MCC, per-class PR-AUC**. Plot **normalized confusion matrix**.
* Tune **class-specific decision thresholds** on the validation PR curves. Calibrate probs with **isotonic/Platt** if needed.

---

### 5) Feature engineering that helps with false positives

Add physics-informed features to separate eclipsing binaries/systematics from planets:

* **Transit shape**: U-shape vs V-shape metrics, duration/period consistency, ingress/egress symmetry.
* **Odd–even depth ratio**, **secondary eclipse depth**.
* **Centroid motion** during transit, **in- vs out-of-transit RMS**, crowding metrics.
* **Multiple event consistency** (period stability via BLS), **S/N**, **harmonics**.
  These usually boost recall for **FALSE_POSITIVE** without crushing generalization.

---

### 6) Training recipe (quick playbook)

1. **Leakage-safe splits:** StratifiedGroupKFold by target ID (k=5).
2. **Weights:** `sample_weight` via inverse frequency.
3. **Search:** Bayesian/Optuna over `max_depth, min_child_weight, subsample, colsample_bytree, gamma, reg_lambda, learning_rate, n_estimators`. Optimize **macro-F1**.
4. **Light resampling:** Try majority undersample 0.6–0.9 + minority oversample 1.0–1.5 (tabular only).
5. **Thresholding:** pick per-class thresholds from val PR curves to reach target recalls on minority classes.
6. **Calibration:** isotonic on a held-out fold.
7. **Report:** macro-F1/MCC + per-class PR-AUC with CIs across folds.

---

### 7) (Optional) Architecture tweaks

* **Two-stage hierarchy:** (A) Planet vs Not-Planet, then (B) Candidate vs False Positive. Different costs per stage.
* **Anomaly detection** for the rarest class to flag hard negatives for review.
* **Ensembles:** blend OVR XGBoost + a small CNN/Transformer on phase-folded light-curve images/embeddings.

---

### 8) If you want, I can help tune it

If you share **class counts per split**, **feature list**, and a **leakage-safe CV setup**, I can sketch exact `sample_weight` values, an Optuna search space, and threshold targets for each class.

**TL;DR:** Skip SMOTE on raw light curves; start with **class/sample weights + leakage-safe CV + macro-F1 objective + per-class thresholds**, and add **astrophysical features**. That combo typically moves minority-class recall the most with minimal hit to overall generalization.

**Comment by prabej7** [comment]
Thank you all @ivenzor , @srinivas-sme , and @arpanghoshal for your valuable guidance.

I applied the suggested changes, especially using class/sample weights, stratified group splits, and feature refinements. The XGBoost model now performs much better on the rare classes while maintaining strong overall generalization.

Your detailed explanations and domain-specific advice made a real difference. I appreciate the time and clarity you all shared. Thank you again for the support!

Once my project is deployed. I would love to post that too!
Thank you :)

**Reply by arpanghoshal** [reply]
All the best!
